### PR TITLE
support coredns v1.11.0

### PR DESCRIPTION
**Pull request by tukwila**
Now coredns release new version: https://github.com/coredns/coredns/releases/tag/v1.11.0, so do upgrade in corefile-migration.
```# nerdctl pull coredns/coredns:1.11.0
docker.io/coredns/coredns:1.11.0:                                                 resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:cc3ebb05fbdba439d2d69813f162aa204b027098c8244fb3156e6e7c0f31c548:    done           |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:28cccdd25b4d971951d50fd6e08ab5fe9a894d04db03557f2713bdf0347e7ddc: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:321c7cb4ac8b4bc06b4e71e1b2526972245292a8502dce49d98843a271b3ba20:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:caa886227134d11f25a4cf824034a597eb7ec611bccdfc2785f7f342ed3e2c27:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:b28ed2365b9cb98936f0420af9de5343264762eb3e20fed4989e22149176664d:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 22.6s                                                                    total:  122.5  (5.4 KiB/s) 
```

And some plugin changes please review if it is right in versions.go?
@chrisohaver @miekg

image

### CODE REVIEW

**Comment by tukwila** [comment]
@chrisohaver Could you please review this PR?

**Comment by tukwila** [comment]
> The plugins map is not correct. It adds plugins that are not supported by the migration, and is missing plugins that it does support. If any new options have been added to supported plugins, those will need to be added.

Do you mean i need not to add any plugins currently? so i modify plugin definition like this,
image
Am i right?

**Comment by chrisohaver** [comment]
I don’t know how to interpret your marked up screen shot. Please take a look at some of the prior PRs in this repo that have added a support for a new coredns version to see how they are done.

Please look at the release notes for the list of plugins the migration package supports. I would not recommend adding to that list at this time.

**Comment by tukwila** [comment]
> I don’t know how to interpret your marked up screen shot. Please take a look at some of the prior PRs in this repo that have added a support for a new coredns version to see how they are done.
> 
> Please look at the release notes for the list of plugins the migration package supports. I would not recommend adding to that list at this time.

Yes, i read  prior plugins and coredns release-notes, then append plugins_1_11_0 cause's in v1.11.0 release-notes there are two new function added into plugins:
plugin/dnstap: add tls support (https://github.com/coredns/coredns/pull/5917)
plugin/k8s_external: add fallthrough option (https://github.com/coredns/coredns/pull/5959)

so, please review my code again. Big thanks.

File: migration/versions.go [hunk 1]
```diff
@@ -745,6 +751,26 @@ var Versions = map[string]release{
...
+	"prometheus":   {},
+	"forward":      plugins["forward"]["v3"],
+	"cache":        plugins["cache"]["v1"],
+	"loop":         {},
+	"reload":       {},
+	"loadbalance":  {},
+	"hosts":        plugins["hosts"]["v1"],
+	"rewrite":      plugins["rewrite"]["v2"],
+	"transfer":     plugins["transfer"]["v1"],
+	"dnstap":  		plugins["dnstap"]["v1"], // add tls support
```
Comment (chrisohaver) [inline migration/versions.go:771]: _dnstap_ is not supported by the migration package at this time.

If you want to add support for it, you'll need to identify in which version of coredns it was introduced, and retroactively add support for all version migrations though 1.11.0.

File: migration/versions.go [hunk 2]
```diff
@@ -745,6 +751,26 @@ var Versions = map[string]release{
...
 }
 
+var plugins_1_11_0 = map[string]plugin{
+	"errors":       plugins["errors"]["v3"],
+	"log":          plugins["log"]["v1"],
+	"health":       plugins["health"]["v1"],
+	"ready":        {},
+	"autopath":     {},
+	"kubernetes":   plugins["kubernetes"]["v8"],
+	"k8s_external": plugins["k8s_external"]["v2"], //add fallthrough option
```
Comment (chrisohaver) [inline migration/versions.go:761]: `plugins["k8s_external"]["v2"]` doesn't exist.  You need to add it to `plugins.go`

File: migration/versions.go [hunk 3]
```diff
@@ -745,6 +751,26 @@ var Versions = map[string]release{
...
+	"prometheus":   {},
+	"forward":      plugins["forward"]["v3"],
+	"cache":        plugins["cache"]["v1"],
+	"loop":         {},
+	"reload":       {},
+	"loadbalance":  {},
+	"hosts":        plugins["hosts"]["v1"],
+	"rewrite":      plugins["rewrite"]["v2"],
```
Comment (chrisohaver) [inline migration/versions.go:769]: _rewrite_ had a new option added to it: https://github.com/coredns/coredns/pull/6004

**Review by chrisohaver — CHANGES_REQUESTED** [review]

File: migration/plugins.go [hunk 4]
```diff
@@ -406,6 +417,13 @@ var plugins = map[string]map[string]plugin{
 				"ttl":  {},
 			},
 		},
+		"v2": plugin{
+			namedOptions: map[string]option{
+				"apex": {},
+				"ttl":  {},
+				"Fall": {},
```
Comment (chrisohaver) [inline migration/plugins.go:424]:
```suggestion
				"fallthrough": {}, // new option
```

File: migration/plugins.go [hunk 5]
```diff
@@ -309,6 +309,17 @@ var plugins = map[string]map[string]plugin{
...
 			},
 		},
+		"v3": plugin{
+			namedOptions: map[string]option{
+				"type":        {},
+				"class":       {},
+				"name":        {},
+				"answer name": {},
+				"edns0":       {},
+				"ttl":         {}, // new option
```
Comment (chrisohaver) [inline migration/plugins.go:319]:
```suggestion
				"ttl":         {},
```

File: migration/versions.go [hunk 6]
```diff
@@ -745,6 +751,25 @@ var Versions = map[string]release{
...
+	"kubernetes":   plugins["kubernetes"]["v8"],
+	"k8s_external": plugins["k8s_external"]["v2"], //add fallthrough option
+	"prometheus":   {},
+	"forward":      plugins["forward"]["v3"],
+	"cache":        plugins["cache"]["v1"],
+	"loop":         {},
+	"reload":       {},
+	"loadbalance":  {},
+	"hosts":        plugins["hosts"]["v1"],
+	"rewrite":      plugins["rewrite"]["v3"],
```
Comment (chrisohaver) [inline migration/versions.go:769]:
```suggestion
	"rewrite":      plugins["rewrite"]["v3"], // add cname_target option
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Thanks! Looks good! Just some minor changes requested on comments.

**Review by chrisohaver — CHANGES_REQUESTED** [review]
There was also a new option added to the cache plugin: `keepttl` https://github.com/coredns/coredns/pull/5879

**Comment by tukwila** [comment]
> ~There was also a new option added to the cache plugin: `keepttl` [coredns/coredns#5879](https://github.com/coredns/coredns/pull/5879)~
> 
> Nevermind - that was added in 1.10.1 ...
> 
> ... Although it looks like it's missing from the 1.10.1 migration. Eventually that should be fixed, but OK if you don't want to do it in this PR.

yes，i review coredns v1.10.1 releasenotes and append cache plugin new added option keepttl. Please review it. -:)

**Comment by tukwila** [comment]
@chrisohaver Could you review my commit again? i hope my commit is within your expected. :)

File: migration/plugins.go [hunk 7]
```diff
@@ -406,6 +426,13 @@ var plugins = map[string]map[string]plugin{
 				"ttl":  {},
 			},
 		},
+		"v2": plugin{
+			namedOptions: map[string]option{
+				"apex": {},
+				"ttl":  {},
+				"Fall": {}, // new option
```
Comment (chrisohaver) [inline migration/plugins.go:433]:
```suggestion
				"fallthtrough": {}, // new option
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Looks good! Thanks!  Just one fix required to the name of the new `fallthrough` option in _k8s_external_.

File: corefile-tool/cmd/validversions_test.go [hunk 8]
```diff
@@ -15,7 +15,7 @@ func TestNewValidVersionsCmd(t *testing.T) {
 		{
 			name: "Works without error",
 			expectedOutput: `The following are valid CoreDNS versions:
-1.1.3, 1.1.4, 1.2.0, 1.2.1, 1.2.2, 1.2.3, 1.2.4, 1.2.5, 1.2.6, 1.3.0, 1.3.1, 1.4.0, 1.5.0, 1.5.1, 1.5.2, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.6.5, 1.6.6, 1.6.7, 1.6.9, 1.7.0, 1.7.1, 1.8.0, 1.8.3, 1.8.4, 1.8.5, 1.8.6, 1.8.7, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.9.4, 1.10.0, 1.10.1
+1.1.3, 1.1.4, 1.2.0, 1.2.1, 1.2.2, 1.2.3, 1.2.4, 1.2.5, 1.2.6, 1.3.0, 1.3.1, 1.4.0, 1.5.0, 1.5.1, 1.5.2, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.6.5, 1.6.6, 1.6.7, 1.6.9, 1.7.0, 1.7.1, 1.8.0, 1.8.3, 1.8.4, 1.8.5, 1.8.6, 1.8.7, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.9.4, 1.10.0, 1.10.1, v1.11.0
```
Comment (chrisohaver) [inline corefile-tool/cmd/validversions_test.go:18]:
```suggestion
1.1.3, 1.1.4, 1.2.0, 1.2.1, 1.2.2, 1.2.3, 1.2.4, 1.2.5, 1.2.6, 1.3.0, 1.3.1, 1.4.0, 1.5.0, 1.5.1, 1.5.2, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.6.4, 1.6.5, 1.6.6, 1.6.7, 1.6.9, 1.7.0, 1.7.1, 1.8.0, 1.8.3, 1.8.4, 1.8.5, 1.8.6, 1.8.7, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.9.4, 1.10.0, 1.10.1, 1.11.0
```

**Comment by tukwila** [comment]
> ... and also fix to version string in valid versions test.

@chrisohaver yes, it's my fault and I corrected. I hope you are online now then review it again. Cross my finger to pray that it can be ok this time.

**Review by chrisohaver — APPROVED** [review]

**Pull request merged**